# Obrada informacija

## Laboratorijska vježba 2

U ovoj vježbi upoznat ćete se s jednom primjenom tehnika obrade informacija u bioinformatici. Ova laboratorijska vježba nosi 4 boda. Izvješće s ove laboratorijske vježbe potrebno je predati u .pdf formatu na *Moodle*. Izvješće koje predajete se mora zvati *PrezimeIme.pdf*. 

Osim biblioteka za rad s Fourierovom transformacijom (koristit ćemo samo numpy) koristit ćemo i biblioteku biopython koja sadrži puno korisnih alata iz područja bioinformatike. Mi ćemo je koristiti za jednostavnije baratanje bioinformatičkim tipovima podataka.

Biblioteka biopython dolazi s instalacijom Anaconde, ali ju je potrebno uključiti u okolinu (*environment*) koja se koristi.

Ako vježbu izvodite u Google Colab okruženju, morate instalirati biblioteku biopython. Instalaciju je potrebno izvršiti u sklopu prvog zadatka ove laboratorijske vježbe. Instalaciju izvodite sljedećim kodom: 

    try:
      import google.colab
      !pip install biopython
    except ImportError:
      pass

Nakon izvođenja ovog koda, možete učitati biopython biblioteku. 

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Zadatak

Python biblioteke potrebne za laboratorijske vježbe su numpy i biopython. Uključite ih ("importirajte") i ispišite verziju svake od njih pomoću \[ime_biblioteke\].__version__.

UPUTA: Osnovna biopython biblioteka ima naziv Bio.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod

!pip install Biopython
import numpy as np
import Bio

print()
print("Numpy verzija:", np.__version__)
print("Biopython verzija:", Bio.__version__)

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/

Numpy verzija: 1.21.6
Biopython verzija: 1.79


2. Zadatak

Uz laboratorijske vježbe dobili ste dvije datoteke s podacima. Datoteku koja sadrži referentni genom jednog soja bakterije  Escherichia coli (escherichia_coli_reference.fasta) u FASTA formatu i datoteku koja sadrži skup očitanja dobivenih sekvenciranjem (ecoli_ILL_small.fastq) u FASTQ formatu.

Datoteke možete učitati koristeći metodu _parse()_ iz biblioteke Bio.SeqIO. Metoda _parse()_ vraća iterator koji možete pretvoriti u Python listu na sljedeći način:

    reads = list(parse("ime_datoteke", "tip_datoteke"))

Tip datoteke postavite na "fasta" ili "fastq".

Učitajte obje datoteke te ispišite broj zapisa u svakoj od njih (broj elemenata u listi). Datoteka koja sadrži referencu trebala bi imati samo jedan zapis, dok bi datoteka s očitanjima trebala sadržavati veći broj zapisa.

NAPOMENA: Ako niste sigurni kako pronaći datoteke na disku iz Jupyter notebook-a, uvijek možete provjeriti radni direktorij sljedećim naredbama:

    import os
    os.getcwd()

i promijeniti ga sa:

    os.chdir()

Ako pak radite u Google Colab okruženju, koristite upute za učitavanje datoteka s Google diska iz prve laboratorijske vježbe. 

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
from Bio.SeqIO import parse

fasta = list(parse("escherichia_coli_reference.fasta", "fasta"))
fastq = list(parse("ecoli_ILL_small.fastq", "fastq"))

print("Broj zapisa u datoteci s referencom:", len(fasta))
print("Broj zapisa u datoteci s očitanjima:", len(fastq))

Broj zapisa u datoteci s referencom: 1
Broj zapisa u datoteci s očitanjima: 38585


3. Zadatak

Svaki zapis koji ste učitali pomoću metode _Bio.SeqIO.parse()_ sadrži Veći broj podataka od kojih su nam bitni samo neki. Naredbom print ispišite cijeli prvi zapis iz datoteke s očitanjima i iz datoteke s referencom.

Vidjet ćete da oba zapisa (među ostalim podacima) sadrže identifikator zapisa i sekvencu. Identifikator zapisa možete dohvatiti pomoću

    zapis.id

dok sekvencu možete dohvatiti pomoću

    zapis.seq

Ispišite identifikator i sekvencu za prvo očitanje te identifikator i prvih 200 znakova za referentni genom E.coli.

NAPOMENA: Referentni genom Escherichia coli je dugačak oko 4.5 milijuna slova

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 
zapis1 = fasta.__getitem__(0)
zapis2 = fastq.__getitem__(0)

print("Prvi zapis iz datoteke s referencom:")
print(zapis1)
print()
print("Prvi zapis iz datoteke s očitanjima:")
print(zapis2)
print()
print("Identifikator za prvo očitanje:")
print(zapis2.id)
print("Sekvenca za prvo očitanje:")
print(zapis2.seq)
print()
print("Identifikator za referentni genom E.coli:")
print(zapis1.id)
print("Prvih 200 znakova za referentni genom E.coli:")
print(zapis1.seq[:200])

Prvi zapis iz datoteke s referencom:
ID: NC_000913.3
Name: NC_000913.3
Description: NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome
Number of features: 0
Seq('AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAG...TTC')

Prvi zapis iz datoteke s očitanjima:
ID: SRR2052522.671
Name: SRR2052522.671
Description: SRR2052522.671 HWI-EAS390_0001:4:1:6915:1123/1
Number of features: 0
Per letter annotation for: phred_quality
Seq('GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAA...TGC')

Identifikator za prvo očitanje:
SRR2052522.671
Sekvenca za prvo očitanje:
GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC

Identifikator za referentni genom E.coli:
NC_000913.3
Prvih 200 znakova za referentni genom E.coli:
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATC

4. Zadatak

Da bismo sekvence DNA analizirali metodama obrade signala, moramo pojednim nukleotidima (slovima) dodijeliti brojčane vrijednosti. Napišite funkciju u Pythonu koja će primiti slovo koje predstavlja nukleotid i vratiti odgovarajuću brojčanu vrijednost. Vrijednosti dodijelite na sljedeći način:

- A =  3
- G =  2
- C = -2
- T = -3

DNA sekvence mogu sadržavati i neke druge znakove (npr. 'N' koji označava da taj nuklotid nije poznat), njima dodijelite vrijednost 0. Također se može dogoditi da nukleotidi budu označeni i malim slovima, pa vodite računa da vaša funkcija mora vratiti ispravnu vrijendost i u tom slučaju.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 

def dohvatiVrijednost(slovo):
  A_value = 3
  G_value = 2
  C_value = -2
  T_value = -3
  N_value = 0

  if slovo == "a" or slovo == "A":
    return A_value
  elif slovo == "g" or slovo == "G":
    return G_value
  elif slovo == "c" or slovo == "C":
    return C_value
  elif slovo == "t" or slovo == "T":
    return T_value
  else:
    return N_value

5. Zadatak

Upotrebite napisanu funkciju da bi od prvog očitanja i od reference kreirali signal. Izračunajte korelaciju pomoću Fourierove transformacije. Zanemarite imaginarne vrijednosti.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 

x1 = [dohvatiVrijednost(slovo) for slovo in zapis1.seq]
x2 = [dohvatiVrijednost(slovo) for slovo in zapis2.seq]

lenRef = len(x1)
lenOci = len(x2)

X1 = np.fft.fft(([0] * lenOci) + x1)
X2 = np.fft.fft(x2 + ([0] * lenRef))

korelacija = np.conjugate(X2)*X1
korelacija = np.fft.ifft(korelacija).real
for i in range(1, len(korelacija)):
  korelacija[i-1] = korelacija[i]
print("Lista svih korelacija:")
print(np.flip(korelacija[:-1]))
print("Od izračunatih", np.size(korelacija)-1, "korelacija najveća korelacija iznosi", np.round(abs(np.max(korelacija))))


Lista svih korelacija:
[ -4. -12.  -9. ...  -1.   2.  -6.]
Od izračunatih 4641772 korelacija najveća korelacija iznosi 693.0


6. Zadatak

Ispišite duljinu reference. Koristeći metode biblioteke _numpy_, izračunajte srednju vrijednost i standardnu devijaciju duljine očitanja (uzmite u obzir sva očitanja).

Primijetit ćete da su sva očitanja jednake duljine.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 
print("Duljina reference:", np.size(zapis1.seq))

duljine_očitanja = [len(očitanje.seq) for očitanje in fastq]
print("Srednja vrijednost duljine očitanja:", np.average(duljine_očitanja))
print("Standardna devijacija duljine očitanja:", np.std(duljine_očitanja))

Duljina reference: 4641652
Srednja vrijednost duljine očitanja: 121.0
Standardna devijacija duljine očitanja: 0.0


7. zadatak

Što ako želimo izračunati korelaciju za veći broj očitanja i istu referencu? To je tipičan slučaj u bioinformatici jer uređaji za sekvenciranje proizvode tisuće i desetke tisuća očitanja koja se potom mapiraju na istu referencu.

Ako korelaciju računamo izravno, potrebno ju je svaki put izračunati iz početka. Ako korelaciju računamo pomoću FFT-a, transformaciju za referencu potrebno je napraviti samo jednom.

Izračunajte korelacije za prvih 10 očitanja.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 

# fft za korelaciju za referencu već imamo u varijabli X1

for j in range(0,10):
  zapis2 = fastq.__getitem__(j)
  x1 = [dohvatiVrijednost(slovo) for slovo in zapis1.seq]
  x2 = [dohvatiVrijednost(slovo) for slovo in zapis2.seq]

  lenRef = len(x1)
  lenOci = len(x2)

  X1 = np.fft.fft(([0] * lenOci) + x1)
  X2 = np.fft.fft(x2 + ([0] * lenRef))

  korelacija = np.conjugate(X2)*X1
  korelacija = np.fft.ifft(korelacija)
  for i in range(1, len(korelacija)):
    korelacija[i-1] = korelacija[i]
  print(j+1,". očitanje", sep='')
  print(np.flip(korelacija[:-1].real))
  print("Maksimalna korelacija", np.ceil(np.max(korelacija).real))
  print()

1. očitanje
[ -4. -12.  -9. ...  -1.   2.  -6.]
Maksimalna korelacija 693.0

2. očitanje
[-6.00000000e+00 -1.30000000e+01 -9.00000000e+00 ... -1.90000000e+01
  7.10545621e-15  9.00000000e+00]
Maksimalna korelacija 711.0

3. očitanje
[ -4. -10. -18. ... -11. -13.  -6.]
Maksimalna korelacija 734.0

4. očitanje
[-6. -5. -9. ... -6. -3.  9.]
Maksimalna korelacija 363.0

5. očitanje
[-6.00000000e+00 -5.00000000e+00  1.00000000e+00 ... -1.60000000e+01
 -4.79617212e-14  9.00000000e+00]
Maksimalna korelacija 663.0

6. očitanje
[ -6. -13. -21. ...   9.  10.   6.]
Maksimalna korelacija 782.0

7. očitanje
[ 4.00000000e+00 -2.13162196e-14 -9.00000000e+00 ...  1.60000000e+01
  5.00000000e+00 -6.00000000e+00]
Maksimalna korelacija 359.0

8. očitanje
[ 6.  5.  9. ... -7. 13.  6.]
Maksimalna korelacija 801.0

9. očitanje
[ -6. -15. -12. ...  -9. -15.  -9.]
Maksimalna korelacija 637.0

10. očitanje
[ -6.  -3.  -6. ...   6. -10.  -6.]
Maksimalna korelacija 677.0



8. zadatak

Na temelju najveće vrijednosti korelacije između reference i prvog očitanja pronađite poziciju na referenci koja je najsličnija očitanju. Pozicija odgovara vrijednosti parametra k za koji je korelacija najveća.

Napišite metodu koja će primiti dva niza znakova jednake duljine, usporediti znakove na istim pozicijama i vratiti broj razlika (Hammingova udaljenost).

"Izrežite" dio reference koji je najsličniji očitanju (iste duljine kao i očitanje) i usporedite ga s očitanjem pomoću napisane funkcije.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 

x1 = [dohvatiVrijednost(slovo) for slovo in zapis1.seq]
x2 = [dohvatiVrijednost(slovo) for slovo in zapis2.seq]

lenRef = len(x1)
lenOci = len(x2)
X1 = np.fft.fft(([0] * lenOci) + x1)
X2 = np.fft.fft(x2 + ([0] * lenRef))

korelacija = np.conjugate(X2)*X1
korelacija = np.fft.ifft(korelacija).real
for i in range(1, len(korelacija)):
  korelacija[i-1] = korelacija[i]
korelacija = np.flip(korelacija)
k_arr = range(-lenOci + 1, lenRef)
k = k_arr[korelacija.argmax()]
indexK = np.where(korelacija == np.max(korelacija))
indexK = list(indexK)
indexK = int(indexK[0])
pocetak = len(korelacija) - k - duljine_očitanja[0] - 120
print("Pozicija koja odgovara vrijednosti parametra k je", indexK-1)
print("Početak za poravnanje je na", pocetak)
zapis2 = fastq.__getitem__(0)
print()
print("Očitanje gornji red / Referenca donji red:")
print(zapis2.seq)
print(zapis1.seq[pocetak : int(duljine_očitanja[0]) + int(pocetak)])

import scipy 
from scipy.spatial.distance import hamming
hamming_distance = len(zapis2.seq) * hamming(zapis2.seq, zapis1.seq[pocetak : int(duljine_očitanja[0]) + pocetak])
print("Hammingova udaljenost je", hamming_distance)

Pozicija koja odgovara vrijednosti parametra k je 2317165
Početak za poravnanje je na 2324486

Očitanje gornji red / Referenca donji red:
GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC
GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGCGCCGCTGCACCATGCCACTCACTGCGCAACATGCGGTGCATATGC
Hammingova udaljenost je 9.0


9. zadatak

U datoteci "ecoli_ILL_small_aln.sam" dana su već izračunata poravnanja svih očitanja na referencu u SAM formatu. SAM je tekstualni "tab separated" format. U prvom stupcu se nalati identifikator očitanja, dok se u četvrtom stupcu nalazi pozicija na referenci na koju je očitanje najbolje poravnato (ostali stupci nas ne zanimaju). Također, datoteka s poravnanjima sadrži i nekoliko _header_ readaka kojima prvi stupac počinje sa znakom '@', njih također možete zanemariti.

Otvorite datoteku s poravnanjima i pronađite poravnanje za prvo očitanje (identifikator očitanja u datoteci s očitanjima i datoteci s poravnanjima mora biti jednak). Usporedite poziciju u datoteci sa pozicijom koju ste dobili pomoću korelacije.

UPOUTA: TSV datoteke možete otvoriti na sljedeći način:

    tsv_file = open("file_name")
    tsv_rows = csv.reader(tsv_file, delimiter="\t")

Varijabla _tsv\_rows_ će sadržavati listu redaka, a svaki redak biti lista vrijednosti (po jedna za svaki stupac).

In [ ]:
from Bio.AlignIO.PhylipIO import sanitize_name
# Ovo je mjesto na kojem možete izvoditi svoj kod. 
import csv

def provjeri(znak):
  if znak == "\t":
    znak = ' '
    return znak
  else:
    return znak

lista_zapisa = [sorted]
with open("ecoli_ILL_small_aln.sam", 'r') as csvfile:
    sam = csv.reader(csvfile, delimiter=' ', quotechar='|')
    for row in sam:
        zapis = ''.join(row)
        if(zapis[0] == 'S'):
          print("Zapis ua prvo očitanje je:")
          print(zapis)
          zapis = list(zapis)
          slozeno = []
          [slozeno.append(provjeri(znak)) for znak in zapis]
          slozeno = ''.join(slozeno)
          slozeno = slozeno.split(" ")
          print()
          print("Identifikator očitanje je", slozeno[0])
          print("Poravnanje za prvo očitanje je", slozeno[3])
          break

Zapis ua prvo očitanje je:
SRR2052522.671	0	NC_000913.3	2324487	60	112M9S	*	0	0	GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC	BGGG?D:GGDD=BFFBFFDEGDGGDDGGDDDEBBED8DDF2??==GEDGDG<<>BGG>G>=A,A=8A<?1=???A##############################################	NM:i:6	ms:i:188	AS:i:188	nn:i:0	tp:A:P	cm:i:13	s1:i:78	s2:i:0	dv:f:0.0320

Identifikator očitanje je SRR2052522.671
Poravnanje za prvo očitanje je 2324487


10. zadatak

Za prvo očitanje pozicija dobivena pomoću korelacije trabala bi biti 2324486, dok je pozicija u datoteci s poravnanjima 2324487. Razlikuju se samo za 1 pa možemo zaključiti da nam je korelacija dala dobru poziciju za poravnanje.

Prisjetimo se da korelacija ne računa točno poravnanje već ju koristimo samo da bi našli kandidatne pozicije za točno računanje. Tek onda na takvim pozicijama možemo točno poravnanje izračunati pomoću algoritama dinamičkog programiranja. Ako bi primijenili dinamičko programiranje za računanje poravnanja očitanja s cijelom referencom, postupak bi bio znatno sporiji i zahtijevao bi veliku količinu radne memorije.

**Ako želite** to možete isprobati pomoću algoritama za poravnanje u biblioteci _bioparser_. Lokalno poravnanje možete izračunati metodom:

    Bio.pairwise2.align.localxx(seq1, seq2)

Za prvih 100 očitanja izračunajte korelaciju te pomoću korelacije poziciju najveće sličnosti očitanja i reference. Usporedite rezultat sa podacima u datoteci s poravnanjima. Ispišite broj očitanja za koja se dvije pozicije razlikuju za najviše 5 mjesta.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod. 
from Bio import pairwise2

print("Lokalno poravnanje:")
for x in pairwise2.align.localxx(zapis2.seq, zapis1.seq[pocetak : int(duljine_očitanja[0]) + pocetak + 10])[:1]:
  print(pairwise2.format_alignment(*x))


Lokalno poravnanje:
1 GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGG-GCCT-CTGT-ATC-ATGCCACTCACTGCGCAAT-ATC-CGGAT-CAA-ATG-------C
  |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| | |||  |||  | | ||||||||||||||||||  ||  ||| | | | |||       |
1 GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTT-GCGCC-GCTG-CA-CCATGCCACTCACTGCGCAA-CAT-GCGG-TGC-ATATGCTGGTTAC
  Score=113



In [ ]:
polje_korelacija = []
def izracunaj_korelaciju():
  x1 = [dohvatiVrijednost(slovo) for slovo in zapis1.seq]
  for index in range(0,100):
    zapis2 = fastq.__getitem__(index)
    x2 = [dohvatiVrijednost(slovo) for slovo in zapis2.seq]

    lenRef = len(x1)
    lenOci = len(x2)
    X1 = np.fft.fft(([0] * lenOci) + x1)
    X2 = np.fft.fft(x2 + ([0] * lenRef))

    korelacija = np.conjugate(X2)*X1
    korelacija = np.fft.ifft(korelacija).real
    for i in range(1, len(korelacija)):
      korelacija[i-1] = korelacija[i]
    korelacija = np.flip(korelacija)
    k_arr = range(-lenOci + 1, lenRef)
    k = k_arr[korelacija.argmax()]

    pocetak = len(korelacija) - k - duljine_očitanja[0] - 120

    polje_korelacija.append(pocetak)

izracunaj_korelaciju()  


In [ ]:
razlikuju = 0

id_lista = []
for i in range(0,100):
  zapis2 = fastq.__getitem__(i)
  id_lista.append(zapis2.id)

with open("ecoli_ILL_small_aln.sam", 'r') as csvfile:
    sam = csv.reader(csvfile, delimiter=' ', quotechar='|')
    counter = 0
    for row in sam:
        zapis = ''.join(row)
        zapis = list(zapis)
        slozeno = []
        [slozeno.append(provjeri(znak)) for znak in zapis]
        slozeno = ''.join(slozeno)
        slozeno = slozeno.split(" ")
        if(slozeno[0] == id_lista[counter] and counter < 100):
          print(counter+1, "- Poravnanje za podatke pod identifikatorom", slozeno[0])
          zapis = list(zapis)
          print("Podatak iz datoteke s poravnanjima", slozeno[3])
          print("Izračunata korelacija za očitanje je", polje_korelacija[counter])
          if (abs(int(slozeno[3]) - int(polje_korelacija[counter])) <= 5):
            razlikuju = razlikuju + 1
          print()
          counter = counter + 1
        if(counter >= 100):
          break

print("Podataka koji se razlikuju za najviše 5 ima:", razlikuju)

1 - Poravnanje za podatke pod identifikatorom SRR2052522.671
Podatak iz datoteke s poravnanjima 2324487
Izračunata korelacija za očitanje je 2324486

2 - Poravnanje za podatke pod identifikatorom SRR2052522.1445
Podatak iz datoteke s poravnanjima 1877261
Izračunata korelacija za očitanje je 1877260

3 - Poravnanje za podatke pod identifikatorom SRR2052522.2358
Podatak iz datoteke s poravnanjima 557778
Izračunata korelacija za očitanje je 557777

4 - Poravnanje za podatke pod identifikatorom SRR2052522.2665
Podatak iz datoteke s poravnanjima 3910489
Izračunata korelacija za očitanje je 1144877

5 - Poravnanje za podatke pod identifikatorom SRR2052522.3419
Podatak iz datoteke s poravnanjima 20240
Izračunata korelacija za očitanje je 3583639

6 - Poravnanje za podatke pod identifikatorom SRR2052522.4013
Podatak iz datoteke s poravnanjima 4051519
Izračunata korelacija za očitanje je 4051518

7 - Poravnanje za podatke pod identifikatorom SRR2052522.5167
Podatak iz datoteke s poravnanjima 28

11. ZAKLJUČAK

Očekivani broj točno pozicioniranih očitanja je 50, jer smo za sada uspješno radili samo s očitanjima na jednom lancu DNA.

Prolaskom kroz zadatke u ovoj vježbi dobili ste kratak uvod u rad s bioinformatičkim podacima i tehnikama obrade signala.